src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs

```diff
@@ -42,116 +42,44 @@ public static bool IsUnnecessaryAsCast(BinaryExpressionSyntax cast, SemanticMode
-            // 1. Dynamic Expressions
-            // 2. If there is any other argument which is dynamic
-            // 3. Dynamic Invocation
-            // 4. Assignment to dynamic
-            if ((expressionType != null &&
-                (expressionType.IsErrorType() ||
-                 expressionType.Kind == SymbolKind.DynamicType)) ||
-                IsDynamicInvocation(castNode, semanticModel, cancellationToken) ||
-                IsDynamicAssignment(castNode, semanticModel, cancellationToken))
-            {
-                return false;
-            }
-
-            if (PointerCastDefinitelyCantBeRemoved(castNode, castedExpressionNode))
-                return false;
+            // Look for simple patterns that are known to be absolutely safe to always remove.
+            if (CastCanDefinitelyBeRemoved(castNode, castedExpressionNode, semanticModel, cancellationToken))
+                return true;
 
-            if (CastPassedToParamsArrayDefinitelyCantBeRemoved(castNode, castType, semanticModel, cancellationToken))
+            // Then look for patterns for cases where we never want to remove casts.  Note: we want these checks to be
+            // very fast, and to eliminate as many cases as necessary.  Importantly, we want to be able to do these
+            // checks before calling into the speculation analyzer.
+            if (CastMustBePreserved(castNode, castedExpressionNode, semanticModel, cancellationToken))
                 return false;
 
-            // A casts to object can always be removed from an expression inside of an interpolation, since it'll be converted to object
-            // in order to call string.Format(...) anyway.
-            if (castType?.SpecialType == SpecialType.System_Object &&
-                castNode.WalkUpParentheses().IsParentKind(SyntaxKind.Interpolation))
-            {
-                return true;
-            }
-
+            // If this changes static semantics (i.e. causes a different overload to be called), then we can't remove it.
             if (speculationAnalyzer.ReplacementChangesSemantics())
                 return false;
 
+            var castTypeInfo = semanticModel.GetTypeInfo(castNode, cancellationToken);
+            var castType = castTypeInfo.Type;
+            var expressionTypeInfo = semanticModel.GetTypeInfo(castedExpressionNode, cancellationToken);
+            var expressionType = expressionTypeInfo.Type;
+
             var expressionToCastType = semanticModel.ClassifyConversion(castNode.SpanStart, castedExpressionNode, castType, isExplicitInSource: true);
             var outerType = GetOuterCastType(castNode, semanticModel, out var parentIsOrAsExpression) ?? castTypeInfo.ConvertedType;
 
-            // Simple case: If the conversion from the inner expression to the cast type is identity,
-            // the cast can be removed.
+            // Clearest case.  We know we haven't changed static semantic, and we have an Identity (i.e. no-impact,
```

Comment:
so there are the runtime semantics of some conversions.  i.e. some conversion actually will execute code at runtime (for example, `(string)obj` or `(int)longVal`).  Other casts have no runtime impact.  like `(object)str`.  

On the other hand, when i say 'static semantics', i mean that the cast's *type* is needed by the compiler to make certain decisions.  For example, this cast has no runtime impact, but is still necessary to determine overloads:

```
Foo((object)null);

// ...

Foo(object o);
Foo(string o);
```

If we didn't have two overloads, then we could remove the cast (because it has no runtime impact, and because the compiler doesn't need the cast to determine the overload). However, if we have both overloads, then the cast needs to stay becaue it' part of determing the semantics of the code *statically*.